base/compiler/types.jl

```diff
@@ -378,6 +378,7 @@ function NativeInterpreter(world::UInt = get_world_counter();
     if world == typemax(UInt)
         world = get_world_counter()
     end
+    world == typemax(UInt) && error("code reflection cannot be used from generated functions")
```

Comment:
@vtjnash This is probably not the best error message for the case that someone uses the abstract interpreter from `@generated`

We should probably instead add a `if get_inference_world(interp) == typemax(UInt)` in the right places.

Reply:
Seems at least fairly accurate though, and the stacktrace should hopefully give the rest of the needed context

Reply:
I moved it into the constructor of InferenceState to guard against other abstract interpreter doing this wrong.